Add embedded clang/llvm for runtime compilation of eBPF programs

brycekahle opened the pull request:
### What does this PR do?

Adds a statically-linked, embedded clang/llvm for runtime compilation of eBPF programs. If enabled, these programs are compiled on-demand. The output object file is cached, such that if any of the following changes, it will be re-compiled:
1. Kernel version
1. Source file hash

Each runtime-compiled source file is pre-processed during CI build to be a concatenation of the `.c` and all user (non-kernel) `.h` files. Also during CI build, a hash is computed for the pre-processed file, such that any modifications on the host would result in refusing to compile the program.

The pre-processed `.c` file is distributed separately from the `system-probe` binary, like our `.o` files, since it must be GPL-licensed to use certain kernel BPF helper functions.

### Motivation

Planned features for networks and/or the system-probe require access to kernel structures that are too difficult to access with offset guessing. Runtime compilation allows us to compile on each host so it is targeted for the running kernel.

### Additional Notes

1. This PR does not address any kind of automatic kernel-header downloading/installation. It will attempt to automatically find compatible headers, but a path can be provided via a config option.
1. The compiler currently does not support in-memory buffers or bundled files. If we decide to keep bundling (currently up for discussion), then that support will be added later.
1. The arm64 system-probe build is currently failing, due to `g++` being killed. I'm investigating why, but it was not happening before I rebased on master.
    1. This is because of a 3GB memory limit on builds. This limit is being revisited.

### Describe your test plan

The kitchen tests currently pass across a wide range of kernels/distros, but the runtime compilations tests only ensure that the programs can be compiled. The CI tests are run another time with the runtime compiler enabled.

Further testing with `system_probe_config.enable_runtime_compiler` turned on would be useful.

brycekahle committed:
Add basic clang 11 compiler

brycekahle committed:
Fix compiler to match local build

brycekahle committed:
Add quotes around KBUILD_MODNAME

brycekahle committed:
Whitespace fixes to compiler C++

brycekahle committed:
Working bundled and non-bundled tests pass

brycekahle committed:
Embedded stdarg.h with compiler

brycekahle committed:
Add config value for runtime compiler output directory

brycekahle committed:
Rename runtime compiler enable setting

brycekahle committed:
Add env var for setting kernel header dirs

brycekahle committed:
Add preprocessor that replaces #include statements with header contents

brycekahle committed:
Fix default BPFDir value to be absolute

brycekahle committed:
Make sure to close non-bundled files that are opened

brycekahle committed:
Add basic content integrity hash generator

brycekahle committed:
Extract compilation logic into runtime asset

brycekahle committed:
Add runtime-compilation of runtime-security probe

brycekahle committed:
Improve output of go generate tools

brycekahle committed:
Fix import cycle

brycekahle committed:
Include content integrity generated files to prevent build problems

`inv deps` did not like it when these files were missing and not generated yet

brycekahle committed:
Working network tracer with runtime compiler

brycekahle committed:
Improve TestEbpfBytesCorrect

brycekahle committed:
Update security probe C path

brycekahle committed:
Add stub for runtime-compiled security probe on unsupported builds

brycekahle committed:
Improve default config BPFDir logic

This figures out the correct path if deployed in a similar repo structure to how it was built. Otherwise, it assumes you are running from the root directory.

brycekahle committed:
Move dst_port to prevent unused variable warning

brycekahle committed:
security-agent doesn't need linux_bpf tag

brycekahle committed:
Use updated sysprobe build images

brycekahle committed:
Fix compiler build on arm64

brycekahle committed:
We don't need terminfo for our use of clang/llvm

brycekahle committed:
Add clang build output tarball to artifacts so we can download it

brycekahle committed:
Use $CI_PROJECT_DIR/.tmp as temp dir

brycekahle committed:
static link libstdc++

brycekahle committed:
wrap select libm (math.h) functions to target older glibc

brycekahle committed:
Move build flags to common function

brycekahle committed:
Install specific version of kernel-headers on CentOS

brycekahle committed:
Fix network tracer build on old kernels

brycekahle committed:
Fix kernel-devel install on older centos versions

brycekahle committed:
Switch to centos 7.6 for system-probe testing

brycekahle committed:
Update buildimage

brycekahle committed:
Extract common functions into headers to be shared between prebuilt/runtime

brycekahle committed:
Remove unused variable in system_probe.py script

hmahmood commented on pkg/ebpf/compiler/shim.cpp:
```diff
@@ -0,0 +1,47 @@
+    auto cppCompiler = (ClangCompiler*) compiler->cpp_compiler;
+    std::vector<const char*> cflags;
+    if (cflagsv) {
+        while(*cflagsv) {
```
Could you get an empty string? Wondering if you need a check like
```c++
while(*cflagsv && **cflagsv != '\0')
```

brycekahle replied:
I don't think an empty string would cause problems.

hmahmood commented on pkg/ebpf/compiler/shim.h:
```diff
@@ -0,0 +1,11 @@
+#ifndef __COMPILER_SHIM_H
+#define __COMPILER_SHIM_H
+
```
Do you need 
```c++
#ifdef __cplusplus
extern "C" {
#endif
```

brycekahle replied:
This is included within the `extern "C"` block in the `.cpp` file, but this couldn't hurt.

brycekahle committed:
Limit compiler stub for security to just linux

brycekahle committed:
Add argument validation to compiler c++

ogaca-dd commented on pkg/ebpf/compiler/compiler.cpp:
```diff
@@ -0,0 +1,273 @@
+std::map<std::string, std::unique_ptr<llvm::MemoryBuffer>> ClangCompiler::remapped_files_;
+
+ClangCompiler::ClangCompiler(const char *name) :
+    llvmContext(new llvm::LLVMContext),
```
You can use `std::make_unique<llvm::LLVMContext>()` to avoid  a memory leak in case of an exception. See https://herbsutter.com/gotw/_102/ for a detailed explanation.

ogaca-dd commented on pkg/ebpf/compiler/compiler.cpp:
```diff
@@ -0,0 +1,273 @@
+    }),
+    theTriple("bpf")
+{
+    if (!ClangCompiler::llvmInitialized) {
```
Note: This code is not thread safe if 2 instances of `ClangCompiler` are created by different threads, maybe not an issue but it worths mention it.

ogaca-dd commented on pkg/ebpf/compiler/compiler.cpp:
```diff
@@ -0,0 +1,273 @@
+    const std::vector<const char*> &extraCflags,
+    bool verbose)
+{
+    std::vector<const char*> cflags;
+    for (auto it = defaultCflags.begin(); it != defaultCflags.end(); it++)
+        cflags.push_back(*it);
```
Can be replaced by `auto cflags = defaultCflags;`

brycekahle replied:
Will that create a copy of `defaultCflags`?

ogaca-dd commented on pkg/ebpf/compiler/compiler.cpp:
```diff
@@ -0,0 +1,273 @@
+        llvm::errs() << "\n";
+    }
+
+    std::unique_ptr<clang::CompilerInvocation> invocation(new clang::CompilerInvocation);
```
Can be replaced by `auto invocation = std::make_unique<clang::CompilerInvocation>()`

ogaca-dd commented on pkg/ebpf/compiler/compiler.cpp:
```diff
@@ -0,0 +1,273 @@
+
+int ClangCompiler::bytecodeToObjectFile(llvm::Module *module, const char *outputFile)
+{
+    if (!module || !outputFile) {
```
If `module` should not be null can it be a reference? `int ClangCompiler::bytecodeToObjectFile(llvm::Module &module, const char *outputFile)`

brycekahle committed:
Fix CGO_LDFLAGS_ALLOW to not clobber on win32

brycekahle committed:
Add missing check for socat in test setup script

ogaca-dd commented on pkg/ebpf/compiler/shim.cpp:
```diff
@@ -0,0 +1,58 @@
+        return -1;
+    }
+
+    auto cppCompiler = (ClangCompiler*) compiler->cpp_compiler;
```
C style cast should be avoided in C++ as there are safer alternatives (static_cast, reinterpret_cast, const_cast).
`static_cast` may works in your case.

ogaca-dd commented on pkg/ebpf/compiler/shim.cpp:
```diff
@@ -0,0 +1,58 @@
+            cflagsv++;
+        }
+    }
+    auto module = std::move(cppCompiler->compileToBytecode(input_file, NULL, cflags, bool(verbose)));
```
`std::move` is not needed here

brycekahle committed:
Merge branch 'master' into bryce.kahle/runtime-compilation

ogaca-dd commented on pkg/ebpf/compiler/compiler.cpp:
```diff
@@ -0,0 +1,273 @@
+
+enum Architecture { PPC, PPCLE, S390X, ARM64, X86 };
+
+std::map<std::string, std::unique_ptr<llvm::MemoryBuffer>> ClangCompiler::remapped_files_;
```
You can use an unnamed namespace for `remapped_files_ ` to avoid name collision.

ogaca-dd reviewed (commented):
I did a quick review on the C++ code: 👍 
Just few minor comments, fell free to ignore them.

brycekahle committed:
improve compiler c++ further

brycekahle committed:
Make LLVM init thread-safe using std::call_once

brycekahle committed:
Add comment clarifying behavior of include_headers

brycekahle commented:
@ogaca-dd Thanks for the C++ review. I made changes in response, if you wanted to take another look.

leeavital commented on pkg/collector/corechecks/ebpf/c/bcc/tcp-queue-length-kern.c:
```diff
@@ -1,5 +1,5 @@
 #include <linux/kconfig.h>
-#define KBUILD_MODNAME "foo"
+#define KBUILD_MODNAME "ddsysprobe"
```
Did this need to change, or is this cleanup ?

leeavital commented on pkg/ebpf/c/bpf_endian.h:
```diff
@@ -0,0 +1,113 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
```
Is this file copied from somewhere ?

brycekahle replied:
From `libbpf`, which is part of the kernel.

leeavital commented on pkg/security/probe/compile.go:
```diff
@@ -0,0 +1,25 @@
+)
+
+// TODO change probe.c path to runtime-compilation specific version
+//go:generate go run ../../ebpf/include_headers.go ../ebpf/c/prebuilt/probe.c ../../ebpf/bytecode/build/runtime/runtime-security.c ../ebpf/c ../../ebpf/c
+//go:generate go run ../../ebpf/integrity.go ../../ebpf/bytecode/build/runtime/runtime-security.c ../../ebpf/bytecode/runtime/runtime-security.go runtime
```
nice

leeavital commented on pkg/ebpf/compiler/compiler.cpp:
```diff
@@ -0,0 +1,273 @@
+    errStream(errString),
+    textDiagnosticPrinter(new clang::TextDiagnosticPrinter(errStream, diagOpts.get())),
+    diagnosticsEngine(new clang::DiagnosticsEngine(diagID, diagOpts, textDiagnosticPrinter.get(), false)),
+    defaultCflags({
```
should we include `KBUILD_MODNAME` here instead of re-defining it in each header

brycekahle replied:
It is defined in `compiler.go`. The other places it is defined, are the BCC compiled probes.

leeavital commented on pkg/ebpf/compiler/compiler.cpp:
```diff
@@ -0,0 +1,273 @@
+        "-fno-asynchronous-unwind-tables",
+        "-fno-stack-protector",
+        "-nostdinc",
+        "-isystem/virtual/lib/clang/include",
```
is this a 'real' path ?

brycekahle replied:
No, hence the `virtual` at the beginning. This is how the embedded files (like `stdarg.h`) get used.

leeavital commented on pkg/ebpf/compiler/compiler.cpp:
```diff
@@ -0,0 +1,273 @@
+
+    if (verbose) {
+        llvm::errs() << "clang invocation:\n";
+        jobs.Print(llvm::errs(), "\n", true);
```
Does this send compilation errors straight to stderr?

brycekahle replied:
It is supposed to send them to the `errString` variable, but it doesn't seem to be working correctly. This is something I need to look into.

brycekahle replied:
Fixed.

ogaca-dd reviewed (commented):
LGTM on the C++ side!

p-lambert commented on pkg/ebpf/runtime_asset.go:
```diff
@@ -0,0 +1,89 @@
+	}
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return "", "", fmt.Errorf("error hashing file %s: %w", f.Name(), err)
+	}
+	if fmt.Sprintf("%x", h.Sum(nil)) != a.hash {
+		return "", "", fmt.Errorf("file content hash does not match expected value")
+	}
+	return p, a.hash, nil
```
nit: should this be extracted to helper function and used by `integrity.go`?

brycekahle replied:
You generally don't want programs run by `go:generate` to import from the same tree (if you can help it) as the packages that use it. Otherwise you have a circular loop since `go:generate` is intended to be part of the build.

p-lambert commented on pkg/network/tracer/tracer.go:
```diff
@@ -143,39 +149,50 @@ func NewTracer(config *config.Config) (*Tracer, error) {
+	if config.EnableRuntimeCompiler {
+		buf, err = getRuntimeCompiledTracer(config)
+		if err != nil {
+			log.Warnf("error compiling network tracer, falling back to pre-compiled: %s", err)
```
Besides logging we probably want a metric reporting how many tracers are running with runtime compilation

brycekahle replied:
Can you be more specific? Do you mean an expvar metric, or something more involved?

p-lambert replied:
IMO it should be one of these telemetry bits we embed in the protobuf messages
cc @leeavital

brycekahle replied:
Is that something we want in this PR, or a separate PR later?

p-lambert replied:
I think it's totally fine to leave to a separate PR as long as we have it before rolling out to customers

p-lambert commented on pkg/network/ebpf/c/runtime/tracer.c:
```diff
@@ -0,0 +1,943 @@
+    }
+}
+
+static __always_inline void increment_telemetry_count(enum telemetry_counter counter_name) {
```
Thoughts on extracting the common inlined functions to something like `common.h`? There is a lot of duplication now between `runtime/tracer.c` and `prebuilt/tracer.c`.

brycekahle replied:
As long as the functions don't require any kind of offset, then yeah we can share them. I would prefer a filename more meaningful than `common.h` though. I've done this a little bit with the HTTP code you added. I didn't want to bloat this PR too much with changes that are tangential to runtime compilation.

p-lambert replied:
Yeah, maybe `tcp.h`, `udp.h` I dunno. I suggested `common.h` as a starting point so you don't have to think too much about this.

p-lambert commented on pkg/ebpf/compiler/compiler.go:
```diff
@@ -0,0 +1,163 @@
+		compiler: C.new_bpf_compiler(),
+		verbose:  verbose,
+	}
+	if err := ebpfCompiler.getErrors(); err != nil {
```
(Ideally the full compilation log should be written to file that can be sent with the agent flare, but this could be done in another PR)

brycekahle replied:
compile errors are now properly ending up in the `err` variable.

brycekahle commented on pkg/ebpf/compiler/compiler.go:
```diff
@@ -0,0 +1,163 @@
+		e.Close()
+	})
+
+	cflags := []string{
+		"-DCONFIG_64BIT",
+		"-D__BPF_TRACING__",
+		`-DKBUILD_MODNAME='"ddsysprobe"'`,
+		"-Wno-unused-value",
+		"-Wno-pointer-sign",
+		"-Wno-compare-distinct-pointer-types",
+		"-Wunused",
+		"-Wall",
+		"-Werror",
+	}
```
Are you saying they should be a package-level variable?

brycekahle commented on pkg/ebpf/config.go:
```diff
@@ -21,14 +27,72 @@ type Config struct {
+	RuntimeCompilerOutputDir string
+}
+
+func curDir() (string, error) {
```
Yes, since the tests don't have access to the configured value (the file/env processing is not run).

brycekahle committed:
Remove remapped_files into ClangCompiler

brycekahle committed:
Use std::make_unique where possible

brycekahle committed:
Remove more C-style casts

brycekahle committed:
Fix compiler output to flow through diagnostics engine

brycekahle committed:
Move default cflags to package-level variable

brycekahle committed:
Add comment about curDir function

brycekahle commented on .gitlab-ci.yml:
```diff
@@ -689,43 +689,49 @@ dogstatsd_arm64_size_test:
   tags: ["runner:docker-arm", "platform:arm64"]
   variables:
     ARCH: arm64
+    PARALLEL_JOBS: 2
```
It was discovered that arm64 builds are being arbitrarily limited to 3GB of memory, so this change may no longer be necessary once that is fixed.

brycekahle committed:
No need for cast in compiler delete

brycekahle committed:
Merge branch 'master' into bryce.kahle/runtime-compilation

brycekahle committed:
Close compiler when done with it

brycekahle committed:
Fix defaultflag usage

brycekahle committed:
use llvm::IntrusiveRefCntPtr for diagnosticsEngine to prevent segfault

brycekahle committed:
Add release note for runtime compilation

brycekahle committed:
Prevent config warning on startup if kernel_header_dirs not set

lebauce commented on pkg/ebpf/integrity.go:
```diff
@@ -0,0 +1,105 @@
+// +build ignore
```
nit: could probably be moved to its own directory

lebauce commented on pkg/ebpf/include_headers.go:
```diff
@@ -0,0 +1,126 @@
+// +build ignore
```
nit: maybe could be moved to its own directory

lebauce reviewed (approved):
Awesome work. I would have preferred if the `compiler` package was in a dedicated Datadog repository so that it could benefit others but I understand this is more convenient.

Gui774ume reviewed (approved):
It's happening !!!! :tada:
Looks great :)
I left a few comments but nothing major.

brycekahle commented on .gitlab-ci.yml:
```diff
@@ -2193,7 +2201,7 @@ kitchen_centos_sysprobe:
   extends: .kitchen_test_system_probe
   before_script:
     - rsync -azr --delete ./ $SRC_PATH
-    - export TEST_PLATFORMS="centos-77,urn,OpenLogic:CentOS:7.7:7.7.201912090"
+    - export TEST_PLATFORMS="centos-76,urn,OpenLogic:CentOS:7.6:7.6.201909120"
```
CentOS 7.6 is our documented minimum supported version.

brycekahle commented on pkg/ebpf/runtime_asset.go:
```diff
@@ -0,0 +1,90 @@
+
+		if err := comp.CompileToObjectFile(inputFile, outputFile, cflags); err != nil {
+			return nil, fmt.Errorf("failed to compile runtime version of %s: %s", a.filename, err)
+		}
```
`defer comp.Close()` is on line 83.

brycekahle commented on pkg/process/config/yaml_config.go:
```diff
@@ -210,6 +210,14 @@ func (a *AgentConfig) loadSysProbeYamlConfig(path string) error {
+
+	if config.Datadog.IsSet(key(spNS, "runtime_compiler_output_dir")) {
+		a.RuntimeCompilerOutputDir = config.Datadog.GetString(key(spNS, "runtime_compiler_output_dir"))
+	}
```
Yes. Good catch!

brycekahle committed:
Add missing config.SetKnown calls

guedou commented on pkg/ebpf/runtime_asset.go:
```diff
@@ -0,0 +1,90 @@
+	io.Closer
+}
+
+// RuntimeAsset represents an asset that needs it content integrity checked at runtime
```
nit:
s/it/its/g

guedou commented on pkg/ebpf/runtime_asset.go:
```diff
@@ -0,0 +1,90 @@
+		return nil, fmt.Errorf("unable to get kernel version: %w", err)
+	}
+
+	inputFile, hash, err := a.Verify(config.BPFDir)
```
I think that there is a [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issue. First, the hash is checked, then the file is opened again and compiled. It leaves an opportunity to modify the file in between these two operations.

brycekahle replied:
You are correct. I will need to modify the compiler to support in-memory buffers rather than files, but we have talked about wanting that ability anyways.

brycekahle replied:
Updated to use in-memory buffers. This should address the TOCTOU issue.

guedou commented on pkg/ebpf/runtime_asset.go:
```diff
@@ -0,0 +1,90 @@
+	// filename includes kernel version and input file hash
+	// this ensures we re-compile when either of the input changes
+	baseName := strings.TrimSuffix(a.filename, filepath.Ext(a.filename))
+	outputFile := filepath.Join(config.RuntimeCompilerOutputDir, fmt.Sprintf("%s-%d-%s.o", baseName, kv, hash))
```
I wonder if we can take advantage of the source files hash embedded in system-probe to compute a signature (for example with HMAC-256) to bind the compiled object to the source file.

brycekahle replied:
Can you elaborate on what you mean here?

guedou replied:
If I am correct, the current code checks the C files hashes before compiling them to eBPF, but will load any object without any check. Using a signature might be useful to check that a resulting object corresponds to an expected C File.

brycekahle replied:
When would the signature be computed and how would it be stored? The object file contents can/will be different on hosts, due to the different kernel versions.

guedou replied:
It could be computed when compiling the object and stored in the filename as currently done with the hash with something like:
`signature  = HMAC(message=CONTENT_OF_THE_OBJECT, key=C_FILE_HASH)`

This is not a strong security feature, but a proposal to check objects before loading them.

brycekahle replied:
One challenge of that approach is currently the expected filename is deterministic from the C file hash. We would need to do some sort of globbing/wildcard approach to find the desired object file.

Since the C files are not embedded in the agent either, I'm not sure a signature check adds much. An attacker could generate an object file with a passing signature. Our currently shipped agent doesn't do any sort of verification of the object files before loading them.

It is worth noting that the system-probe runs as root, and thus these object files would also be owned by root. You would need root access to replace them.

brycekahle committed:
Merge branch 'master' into bryce.kahle/runtime-compilation

brycekahle committed:
Fix comment typo

brycekahle committed:
Update runtime asset hashes

brycekahle committed:
Change to use in-memory buffer for compiler input

xornivore reviewed (approved):
lgtm for container-integrations

brycekahle committed:
Fix potential deadlock in http monitor

brycekahle committed:
Allow compiler to compile files OR buffers

brycekahle committed:
Merge branch 'master' into bryce.kahle/runtime-compilation

sunhay commented on pkg/process/config/config.go:
```diff
@@ -23,14 +23,19 @@ import (
 	defaultSystemProbeBPFDir = "/opt/datadog-agent/embedded/share/system-probe/ebpf"
 
+	// defaultRuntimeCompilerOutputDir is the default path for output from the system-probe runtime compiler
+	defaultRuntimeCompilerOutputDir = "/var/tmp/datadog-agent/system-probe/build"
```
This is also defined separately as a default in `pkg/ebpf/config.go`. Might be worth cleaning up so its defined once. Fine if this is done in a separate follow-up PR .

sunhay reviewed (approved):
LGTM on the Processes-side for configuration. I haven't dug into the actual runtime compilation pieces given its been thoroughly reviewed and approved.

brycekahle committed:
Include hash of cflags in output filename

This ensures we re-compile assets when config options change and/or we update the default flags.

brycekahle committed:
Merge branch 'master' into bryce.kahle/runtime-compilation

brycekahle committed:
Merge branch 'master' into bryce.kahle/runtime-compilation

brycekahle committed:
Merge remote-tracking branch 'origin/master' into bryce.kahle/runtime-compilation

brycekahle committed:
Update runtime assets

brycekahle committed:
Run kitchen tests another time using the runtime compiled probes

This doesn't include DNS or HTTP tests at the moment because some refactoring is needed to make that possible.

brycekahle committed:
Merge remote-tracking branch 'origin/master' into bryce.kahle/runtime-compilation

brycekahle committed:
Passthrough wait_thr to check_output

brycekahle committed:
Make it clear which version of tests are running
